### PR TITLE
Early iterator closing in dispatch

### PR DESCRIPTION
**Pull request by josephschorr**
Add a lint check and fix usages of iterators to ensure they are closed as early as we can

### CODE REVIEW

**Review by vroldanbet — CHANGES_REQUESTED** [review]
Why is the analyzers binary being checked in?

**Comment by josephschorr** [comment]
> Why is the analyzers binary being checked in?

Must have accidentally added it. Removed

File: tools/analyzers/closeafterusagecheck/closeafterusagecheck_test.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"testing"
+
+	"golang.org/x/tools/go/analysis/analysistest"
+)
+
+func TestAnalyzer(t *testing.T) {
+	analyzer := Analyzer()
+	analyzer.Flags.Set("must-be-closed-after-usage-types", "*notimmediatelyclosed.SomeIterator")
```
Comment (vroldanbet) [inline tools/analyzers/closeafterusagecheck/closeafterusagecheck_test.go:11]: unhandled error

**Review by vroldanbet — COMMENTED** [review]
👋🏻 I fail to see how this help with DB connection connection, maybe I'm missing something. The only implementation of a relationship iterator used by production code is `sliceRelationshipIterator`, which does not hold any references to database connections or any state that would cause DB conns to be released.

**Review by vroldanbet — APPROVED** [review]
👍🏻 looks good to me

**Pull request merged**